MSPileup service HTTP APIs

vkuznet opened the pull request:
Fixes #11425 

#### Status
in development

#### Description
The MSPileup service HTTP APIs

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
#11438 

#### External dependencies / deployment changes
The MSPileup HTTP service APIs follow the same MS implementation, i.e. the service is part of MSManager.py. But since it requires support of various HTTP methods, I enhanced MSManager functionality to support `POST, PUT, DELETE` HTTP methods. Due to various constrains of current REST layer, e.g. routes hierarchy, I was forced to introduce `access` end-point to support all HTTP methods. Here is how the end-points look like via curl call:
```
echo "HTTP GET API calls"
curl -v "http://localhost:8249/ms-pileup/data/access?pileupName=pileup"
curl -v "http://localhost:8249/ms-pileup/data/access?campaign=Campaign"
# and we can use filters to extract certain fields from the doc, e.g.
curl -v "http://localhost:8249/ms-pileup/data/access?pileupName=pileup&filters=["campaigns", "pileupType"]"


echo "HTTP POST API calls"
# first create document
curl -v -X POST -H "Content-type: application/json" -d '{"bla":1}' \
       http://localhost:8249/ms-pileup/data/access
# second, query the data via provided JSON query (spec)
curl -v -X POST -H "Content-type: application/json" \
       -d '{"query":{"pileupName":"bla"}}' \
       http://localhost:8249/ms-pileup/data/access
# third, use query and filters
curl -v -X POST -H "Content-type: application/json" \
       -d '{"query":{"pileupName": "bla"}, "filters":["campaigns"]}' \
        http://localhost:8249/ms-pileup/data/access

echo "HTTP PUT API call"
curl -v -X PUT -H "Content-type: application/json" -d '{"bla":1}' \
        http://localhost:8249/ms-pileup/data/access

echo "HTTP DELETE API calls"
curl -v -X DELETE -H "Content-type: application/json" -d '{"bla":1}' \
        http://localhost:8249/ms-pileup/data/access
curl -v -X DELETE -H "Content-type: application/json" -d '{"pileupName":"pileup"}' \
        http://localhost:8249/ms-pileup/data/access
```

Please note, that `POST` method has dual role. It can be used to create resources in MSPileup, or it can be used to fetch data for provided JSON query (spec). In latter case, the query should be constructed as `{'query': ...}` spec, i.e. it has `query` key.

vkuznet committed:
MSPileup service HTTP APIs

vkuznet commented:
@amaltaro , @todor-ivanov this is first draft for HTTP service API, it is not complete and represents work in progress. But it is fully functional in my dev setup. Since it depends on #11438 , I added `MSPileupData` mockup class (which comes from #11438 ) to check the functionality of HTTP service APIs and being able to continue development. At this moment, I need your feedback about general functionality (and I do remember that we'll need to add authz, etc.). First, we need to decide on end-points, and HTTP method access. Please review description of how APIs will be accessed and provide your initial feedback.

vkuznet committed:
Unit test for MSPileup module

vkuznet commented:
test this please

vkuznet commented:
test this please

amaltaro commented:
@vkuznet Valentin, has this PR been updated with the final changes made to the https://github.com/dmwm/WMCore/pull/11438 ? Is it ready for review?

vkuznet commented:
Alan, I rebased the PR, and adjusted code to take data layer changes from #11438, and now you can review this PR. Please note, it is still a draft version since I need initial input about APIs, functionality and most importantly end-points. Once I'll understand how client will communicate with service and we will decide on GET/POST/PUT end-points and parameters, then I can proceed with implementation.

amaltaro commented on src/python/WMCore/MicroService/Service/RestApiHub.py:
```diff
@@ -34,3 +34,5 @@ def __init__(self, app, config, mount):
         self._add({"status": Data(app, self, config, mount),
                    "info": Data(app, self, config, mount)})
+        if 'pileup' in mount:
+            self._add({"access": Data(app, self, config, mount)})
```
I think I asked this question somewhere else. But can you please expand why the `access` endpoint is required?

amaltaro reviewed (commented):
Valentin, I think the way you implemented those 4 methods are fairly complete.
I left a few comments here and there on things that I saw during the review.

On top of that, I think we should define a specific endpoint API for reaching the MSPileup service, instead of using `ms-pileup/data` as suggested in your Data.py docstring. We should define a new API, e.g. `ms-pileup/data/pileup`

vkuznet commented on src/python/WMCore/MicroService/MSCore/MSManager.py:
```diff
@@ -166,6 +179,18 @@ def _parseConfig(self, config):
+        res = self.msPileup.status()
+        endTime = datetime.utcnow()
+        self.updateTimeUTC(res, startTime, endTime)
+        self.logger.info("Total pileup execution time: %d secs", res['execution_time'])
```
done

vkuznet commented on src/python/WMCore/MicroService/MSCore/MSManager.py:
```diff
@@ -77,6 +77,19 @@ def __init__(self, config=None, logger=None):
+            thname = 'MSPileup'
+            self.pileupThread = start_new_thread(thname, daemon,
+                                                 (self.pileup,
+                                                  'assigned',
```
done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,112 @@
+        # for authz we should provide relevant section in MS configuration
+        # if no role/group is provided then authz_match used in HTTP APIs
+        # will always return valid state. Therefore, for test purposes we can
+        # use empty role and group, while for production setup we should set thenm up
```
I suggest to make it last change as we still need to define proper config. So far I put basic `self.role` and `self.group` which get its values from passed config (since I do not know yet how config will look like), and I used them in `authz_match` function. If you'll provide proper config then we can adjust this.

amaltaro replied:
Giving it further thought, I would suggest to create a specific GH issue for the authorization rules and implementation. It's a requirement for deploying this service to production though. Can you please create a ticket for that?

vkuznet commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,112 @@
+
+        :return: status dictionary
+        """
+        summary = dict(OUTPUT_REPORT)
```
I do not know since it depends what will be used by metrics. For instance the status may report status of MSPileup instance along with status of underlying data layer stats. Again, I suggest to make necessary changes once we'll address monitoring issue #11429

vkuznet commented on src/python/WMCore/MicroService/Service/Data.py:
```diff
@@ -35,6 +54,17 @@
 from WMCore.REST.Format import JSONFormat, PrettyJSONFormat
 
 
+def mspileupError(doc):
```
ok, moved

vkuznet commented on src/python/WMCore/MicroService/Service/Data.py:
```diff
@@ -52,6 +82,7 @@ def __init__(self, app, api, config, mount):
         RESTEntity.__init__(self, app, api, config, mount)
         self.config = config
         arr = config.manager.split('.')
+        self.mount = mount
```
we need a way to distinguish which object we're dealing with in order to use proper set of get/post/put/delete APIs. The constructor of `class Data` only uses `app, api, config, mount` parameters. The app is instance of `WMCore.REST.Main.RESTDaemon` , the api is instance of `WMCore.MicroService.Service.RestApiHub.RestApiHub` (where the actual object is present) and the rest available is mount. So far I did not found any way to know which object is passed and I used mount string to define that. If you  have better suggestion please let me know.

vkuznet commented on src/python/WMCore/MicroService/Service/Data.py:
```diff
@@ -89,12 +120,25 @@ def get(self, **kwds):
             res.update(self.mgr.info(kwds.pop("request", None), **kwds))
-        return results(res)
+            yield res
+        elif 'pileup' in self.mount:
```
Well, api here is MSManager instance under ResApiHub, i.e. `api=<WMCore.MicroService.Service.RestApiHub.RestApiHub object at 0x10d6763d0>` since MSPileup is part of MSManager object..

vkuznet commented:
Alan, I adjusted this PR according to your feedback. Please have another look.

amaltaro reviewed (commented):
Valentin, I stopped this review at the MSManager module because I am now confused with these changes and the chain of dependency for the REST layer.

From what I understand, the data flow for a REST call execution is:
1) an income HTTP request arrives at the MicroService/Service/Data module;
2) the method executed in Data.py module goes along with the HTTP method defined;
3) then Data.py module makes a call to MSManager endpoint (method named after the REST API name/endpoint);
4) MSManager deals with it, and if needed, makes another call to the specific MS service module.
5) result is then yield to the user, through other REST layers in WMCore.

If this understanding is accurate, then I think those get/put/post/delete methods should be moved from MSManager.py module over to the Data.py one, no?

vkuznet commented:
Alan, regarding your comment here https://github.com/dmwm/WMCore/pull/11443#pullrequestreview-1278093254 
You need to draw the line of data/request flow and it goes like this:
```
REST Server
    |---> RestApiHub
               |--> Data.py
                         |   (it handles get/post/put/delete HTTP APIs)
                         |--> self.mgr
                                   |  (via config which provides
                                   |       data.manager = 'WMCore.MicroService.MSCore.MSManager.MSManager')
                                   |--> MSManager
                                             |  (it routes to get/post/put/delete if available)
                                             |--> MSxxx (where xx is concrete MS service object)
```
So, since `RestApiHub` talks to MSManager via `Data` class which by itself performs additional stuff like HTTP input parameter validation you need to follow the chain. Otherwise, if you'll need to put logic of `Data` class into MSManager and make it dependent from `RESTEntity`. Moreover, the `RestApiHub` defines end-points which it delegates to `Data` class. I think the separation `RestApiHub -> Data -> MSManager` is good enough. In other words, for every HTTP method (GET/POST/...) the REST server calls `Data` which validates the data and ask its manager to deal with request.

Said that, if you move `get/post/put/delete` methods to MSManager then MSManager scope will be much bigger. So far it manages the MS services, that's it. It does not know about HTTP logic. But if you move these methods to MSManager it will have dual scope: it will manage MS services and their APIs, and it should manage HTTP requests, like validate parameters, handle HTTP errors, etc. I doubt it is what we need, and I think the current implementation correctly split these responsibilities.

vkuznet commented:
Here is new issue for managing auth rules/groups/roles: https://github.com/dmwm/WMCore/issues/11465

amaltaro commented:
Thanks for creating the new issue.

About the data flow, I don't think we contradict each other. However:

> Said that, if you move get/post/put/delete methods to MSManager then MSManager scope will be much bigger. So far it manages the MS services, that's it. It does not know about HTTP logic. 

is not what I suggested in my previous comment. I think we should keep all the HTTP logic in the Data.py module, which is already performing some data sanitization and handling a couple of HTTP methods (get/post).
MSManager - as you correctly pointed out - is dynamically loaded from the configuration and, in addition to managing the microservice daemon itself, it also has an object to the microservice itself, where it can reach the logic for the endpoint APIs.

So, what I am suggesting is to actually define the other 2 HTTP methods in the Data.py module (that would be put/delete), then call the relevant API in MSManager (for now we only have status/info). This is how we could keep things more consistent with the way they are currently implemented. And having the HTTP methods defined in the Data module seem more logical to me.

PS.: As we discuss this, I think in the future we would have to separate MSManager into 2 different modules; one managing the service/daemon itself; and a second to manage the endpoints. Another way would be to define a specific `data.manager` module for each microservice. Just a thought for the future though.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,125 @@
+from WMCore.MicroService.MSPileup.MSPileupData import MSPileupData
+
+
+def mspileupError(doc):
```
Valentin, can you please provide a unit test for this function?

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,125 @@
+        summary.update({'thread_id': current_thread().name})
+        return summary
+
+    def get(self, **kwargs):
```
If it does not increase complexity of the code down the road, I would really like to avoid these method names in this class. Names like `get`, `update` are present in many python objects and it always adds extra confusion when reading the code.
How about renaming it get --> getPileup? I will make similar comment to the methods below.

Keep in mind that MSManager will have to be updated (maybe other places as well).

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,125 @@
+
+        return results
+
+    def query(self, spec, projection=None):
```
How about renaming it to `queryDatabase`? Docstring can be improved as well.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,125 @@
+        """
+        return self.mgr.getPileup(spec, projection)
+
+    def create(self, pdict):
```
How about rename it to `createPileup`? Docstring can be reworded as well

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,125 @@
+        authz_match(self.role, self.group)
+        return self.mgr.createPileup(pdict)
+
+    def update(self, pdict):
```
I'd strongly encourage to rename it to `updatePileup`. Docstring could use of some rewording as well

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -4,14 +4,125 @@
+        authz_match(self.role, self.group)
+        return self.mgr.updatePileup(pdict)
+
+    def delete(self, spec):
```
Lastly, I'd rename it to `deletePileup`. Same comment for the docstring.

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, in addition to some minor (still relevant IMO) comments made along the code, can you please provide a services_config PR with the required changes for MSPileup in `preprod`?
Once we merge this, as we discussed, we want to deploy it in testbed and run more realistic tests.

amaltaro commented on test/python/WMCore_t/MicroService_t/MSPileup_t/MSPileup_t.py:
```diff
@@ -0,0 +1,113 @@
+    def testMSPileupError(self):
+        "test mspileupError function"
+        doc = {'error': 'mspileup error', 'code': 123, 'message': 'msg'}
+        try:
```
Please use a construction like:
```
        with self.assertRaises(cherrypy.HTTPError):
            mspileupError(doc)
```

otherwise we can't be sure whether this test is really testing that function.

In addition, please add a success case as well.

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, other than updating the unit test, I think this PR is ready to proceed to the next phase.
Once you update the unit test (feel free to squash accordingly), please request another review and remove labels that are no longer relevant.

amaltaro commented:
And just a reminder to provide preprod services_config PR.

amaltaro commented:
Please provide the relevant services_config preprod configuration, based on the `test` branch. For the mongo credentials, you can use the same as those provided for ms-output in the preprod branch.

vkuznet commented:
Alan, I added pileup area which includes secrets and config into gitlab preprod branch, see https://gitlab.cern.ch/cmsweb-k8s/services_config/-/tree/preprod/reqmgr2ms-pileup

amaltaro commented on test/python/WMCore_t/MicroService_t/MSPileup_t/MSPileup_t.py:
```diff
@@ -0,0 +1,109 @@
+        with self.assertRaises(cherrypy.HTTPError):
+            mspileupError(doc)
+        success = {}
+        res = mspileupError(success)
```
there is a pylint error on this line. This function has no return. Please fix it in the upcoming PR that you work for for MSPileup.

amaltaro reviewed (approved):
As mentioned along the code, there is one error reported by pylint that needs to be fixed in the unit test. Please fix that in your near future PR. In addition to that, I made this merge request for the service configuration:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/191

and everything else looks good to me. Thank you very much, Valentin.
